lib/ecto/soft_delete_repo.ex

```diff
@@ -81,11 +81,37 @@ defmodule Ecto.SoftDelete.Repo do
         if has_include_deleted_at_clause?(query) || opts[:with_deleted] || !soft_deletable?(query) do
           {query, opts}
         else
-          query = from(x in query, where: is_nil(x.deleted_at))
+          query = filter_soft_deleted(query)
           {query, opts}
         end
       end
 
+      # We need to check the entire query and apply filtering
+      # where appropriate.  So, we recurse the query here and
+      # rebuild it with filtering where appropriate.  This
+      # currently only considers the source and does not handle
+      # things like joins...
+      defp filter_soft_deleted(%Ecto.Query{from: %{source: {_schema, _module}}} = query) do
+        if Ecto.SoftDelete.Query.soft_deletable?(query) do
+          from(x in query, where: is_nil(x.deleted_at))
+        else
+          query
+        end
+      end
+
+      defp filter_soft_deleted(%Ecto.SubQuery{query: query} = sub) do
+        if Ecto.SoftDelete.Query.soft_deletable?(query) do
+          from(x in query, where: is_nil(x.deleted_at)) |> subquery()
```

Comment:
I think this is going to throw away any _other_ stuff that's stored in the subquery's other attributes like `:cache`, `:params`, and `:select` (looking at [this type](https://hexdocs.pm/ecto/Ecto.SubQuery.html#t:t/0) or [this struct](https://github.com/elixir-ecto/ecto/blob/v3.12.4/lib/ecto/query.ex#L11)). But maybe that concern can be addressed by updating the existing subquery instead of creating a new one?

```suggestion
          %{sub | query: from(x in query, where: is_nil(x.deleted_at))}
```

Reply:
Oh, shoot yeah good catch.